File: ejercicios/fase-1/validar-salida-llm-pydantic/compra.py
"""Ejercicio 1.4 — Validar la salida de un LLM con pydantic (Primero-Sin-IA).

Un LLM extrajo una compra desde el texto de un correo y devolvió un JSON. NO confíes
en él: puede mandar el monto como string, dejar campos en blanco, devolver una lista
de items vacía, una fecha que no es fecha, o ALUCINAR campos que no pediste.

Tu trabajo: diseñar el modelo `Compra` (pydantic v2) que valide ese JSON en la FRONTERA
—lo acepta limpio y tipado, o lo rechaza con ValidationError— e implementar `parsear_compra`.

Contrato del modelo (esto es tu spec; los tests lo fijan):
    comercio:   str  no vacío ni de solo espacios   (ej. "Lider")
    monto:      int  > 0, en pesos chilenos          (ej. 12990; acepta "12990" y lo coacciona)
    categoria:  str  no vacío ni de solo espacios     (ej. "supermercado")
    fecha:      date ISO (YYYY-MM-DD)                  (ej. "2026-06-20")
    items:      list[str] con AL MENOS un elemento     (ej. ["leche", "pan"])
    + campos NO declarados (alucinados por el LLM) => la validación debe FALLAR.

Pistas de diseño (no la solución):
  - Constraints con Field: gt, min_length.
  - "   " (solo espacios) pasa min_length: necesitas un @field_validator con strip.
  - Cerrar la puerta a campos extra: model_config = ConfigDict(extra="forbid").
  - parsear + validar en UN paso: Compra.model_validate_json(...).

Instala una vez:  uv add pydantic   (o: pip install pydantic)
Corre los tests:  uv run pytest
"""

from pydantic import BaseModel


class Compra(BaseModel):
    # TODO(estudiante): define los campos, sus constraints y los validadores que
    # haga falta, según el contrato del docstring de arriba. Borra este `pass`.
    pass


def parsear_compra(raw_json: str) -> "Compra":
    """Valida la salida (JSON crudo) de un LLM contra el modelo Compra.

    Debe parsear el string JSON Y validar a la vez (un solo método de pydantic v2).
    Si el JSON no cumple el contrato, deja que se propague el ValidationError.
    """
    # TODO(estudiante): implementa con pydantic v2 (no uses json.loads por separado).
    raise NotImplementedError("Implementa parsear_compra con pydantic v2")

File: ejercicios/fase-1/validar-salida-llm-zod/compra.ts
/**
 * Ejercicio 1.8 B — Validar la salida de un LLM con zod (Primero-Sin-IA).
 *
 * Un LLM extrajo una compra desde un correo y devolvió este JSON. NO confíes en él.
 * Diseña el schema que lo valide EN LA FRONTERA.
 *
 * Reglas del schema (los seis campos):
 *   - comercio:  string, NO vacío ni de solo espacios.
 *   - monto:     entero positivo; debe COACCIONAR "12990" (string) a 12990 (number).
 *   - moneda:    solo "CLP" o "USD".
 *   - categoria: string, NO vacío ni de solo espacios.
 *   - fecha:     string con formato ISO de fecha (YYYY-MM-DD).
 *   - items:     array de strings (no vacíos) con al menos 1 elemento.
 *   - RECHAZA cualquier campo alucinado (extra) que el LLM pueda inventar.
 *
 * Pista de la lección (4.8): `z.object` IGNORA los campos extra. Para rechazarlos,
 * necesitas otra forma de objeto. Empieza con `z.object` y deja que el test de
 * "campos alucinados" te diga que tienes que cambiarla.
 */
import { z } from "zod";

export const CompraSchema = z.object({
  // TODO: define los seis campos con sus constraints.
});

// El tipo Compra se INFIERE del schema. No lo escribas a mano.
export type Compra = z.infer<typeof CompraSchema>;

/** Parsea el JSON crudo Y lo valida en una sola función. Lanza si es inválido. */
export function parsearCompra(rawJson: string): Compra {
  throw new Error("TODO: JSON.parse(rawJson) + validación con zod");
}
